### DISCUSSION TITLE
4.2장 User 레이어 구현 관련 질문

### DISCUSSION THREAD

**Discussion by hyorim-Han** [fsoftwareengineer/todo-application#112] [질문과 답변] [answered]
4.2장 User 레이어 구현 관련 질문 있습니다.

UserEntity에서 책에서 나온 원본 코드 그대로 작성하면 빌드가 되지만, User레이어의 변수명을 바꾸게 되면 오류가 납니다.

예를 들면
책의 변수 앞에 u_를 붙여 빌드하게 되면 findByEmailAndPassword()함수에서 오류가 납니다.
---------------------------------------------------------------------------------------------------------------------
UserEntity.java

`package com.lululala.peopleofwork.model;

import javax.persistence.Column;
import javax.persistence.Entity;
import javax.persistence.GeneratedValue;
import javax.persistence.Id;
import javax.persistence.Table;
import javax.persistence.UniqueConstraint;

import org.hibernate.annotations.GenericGenerator;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

@Data
@Entity
@Builder
@NoArgsConstructor
@AllArgsConstructor
@Table(uniqueConstraints = {@UniqueConstraint(columnNames = "u_email")})

public class UserEntity {
	@Id
	@GeneratedValue(generator="system-uuid")
	@GenericGenerator(name = "system-uuid", strategy = "uuid")
	private String id; //사용자에게 부여되는 고유 id
	
	@Column(nullable = false)
	private String u_name; // 사용자 이름

	@Column(nullable = false)
	private String u_email; // 사용자 이메일(Id)
	
	@Column(nullable = false)
	private String u_password; // 사용자 패스워드
}`
---------------------------------------------------------------------------------------------------------------------

UserService.java

`package com.lululala.peopleofwork.service;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import com.lululala.peopleofwork.model.UserEntity;
import com.lululala.peopleofwork.persistance.UserRepository;

import lombok.extern.slf4j.Slf4j;

@Slf4j
@Service
public class UserService {
	
	@Autowired
	private UserRepository userRepository;
	
	public UserEntity create(final UserEntity userEntity) {
		
		if(userEntity == null || userEntity.getU_email() == null) {
			throw new RuntimeException("Invalid arguments");
		}
		
		final String u_email = userEntity.getU_email();
		if(userRepository.existsByEmail(u_email)) {
			log.warn("Email already exists {}", u_email);
			throw new RuntimeException("Email already exists");
		}
		
		return userRepository.save(userEntity);
	}
	
	public UserEntity getByCredentials(final String u_email, final String u_password) {
		return userRepository.findByEmailAndPassword(u_email, u_password);
	}
}
`
---
UserController.java

`package com.lululala.peopleofwork.controller;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import com.lululala.peopleofwork.dto.ResponseDTO;
import com.lululala.peopleofwork.dto.UserDTO;
import com.lululala.peopleofwork.model.UserEntity;
import com.lululala.peopleofwork.service.UserService;

import lombok.extern.slf4j.Slf4j;

@Slf4j
@RestController
@RequestMapping("/auth")
public class UserController {
	
	@Autowired
	private UserService userService;
	
	@PostMapping("/signup")
	public ResponseEntity<?> registerUser(@RequestBody UserDTO userDTO){
		try {
			// 요청을 이용해 저장할 사용자 만들기
			UserEntity user = UserEntity.builder()
					.u_email(userDTO.getU_email())
					.u_name(userDTO.getU_name())
					.u_password(userDTO.getU_password())
					.build();
			
			// 서비스를 이용해 리포지터리에 사용자 저장
			UserEntity registeredUser = userService.create(user);
			UserDTO responseUserDTO = UserDTO.builder()
					.u_email(registeredUser.getU_email())
					.u_name(registeredUser.getU_name())
					.id(registeredUser.getId())
					.build();
			
			return ResponseEntity.ok().body(responseUserDTO);
		} catch (Exception e) {
			//사용자 정보는 항상 하나이므로 리스트로 만들어야 하는 ResponseDTO를 사용하지 않고 그냥 UserDTO 리턴
			
			ResponseDTO responseDTO = ResponseDTO.builder().error(e.getMessage()).build();
			
			return ResponseEntity.badRequest().body(responseDTO); 
		}
	}
	
	@PostMapping("/signin")
	public ResponseEntity<?> authenticate(@RequestBody UserDTO userDTO){
		UserEntity user = userService.getByCredentials(
				userDTO.getU_email(),
				userDTO.getU_password());
		
		if(user != null) {
			final UserDTO responseUserDTO = UserDTO.builder()
					.u_email(user.getU_email())
					.id(user.getId())
					.build();
			
			return ResponseEntity.ok().body(responseUserDTO);
		} else {
			ResponseDTO responseDTO = ResponseDTO.builder()
					.error("Login failed.")
					.build();
			
			return ResponseEntity.badRequest().body(responseDTO);
		}
	}
}
`
---
UserRepository.java

`package com.lululala.peopleofwork.persistance;

import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.stereotype.Repository;

import com.lululala.peopleofwork.model.UserEntity;

@Repository
public interface UserRepository extends JpaRepository<UserEntity, String> {
	UserEntity findByEmail(String u_email);
	Boolean existsByEmail(String u_email);
	UserEntity findByEmailAndPassword(String u_email, String u_password);
}
`
---
UserDTO.java
`package com.lululala.peopleofwork.dto;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class UserDTO {
	private String token;
	private String u_email;
	private String u_name;
	private String u_password;
	private String id;
}`
---

제가 작성한 코드는 위와 같고 오류는 

Error creating bean with name 'userController': Unsatisfied dependency expressed through field 'userService'; nested exception is org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'userService': Unsatisfied dependency expressed through field 'userRepository'; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'userRepository' defined in com.lululala.peopleofwork.persistance.UserRepository defined in @EnableJpaRepositories declared on JpaRepositoriesRegistrar.EnableJpaRepositoriesConfiguration: Invocation of init method failed; nested exception is org.springframework.data.repository.query.QueryCreationException: Could not create query for public abstract com.lululala.peopleofwork.model.UserEntity com.lululala.peopleofwork.persistance.UserRepository.findByEmailAndPassword(java.lang.String,java.lang.String)! Reason: Failed to create query for method public abstract com.lululala.peopleofwork.model.UserEntity com.lululala.peopleofwork.persistance.UserRepository.findByEmailAndPassword(java.lang.String,java.lang.String)! No property email found for type UserEntity! Did you mean 'u_email'?; nested exception is java.lang.IllegalArgumentException: Failed to create query for method public abstract com.lululala.peopleofwork.model.UserEntity com.lululala.peopleofwork.persistance.UserRepository.findByEmailAndPassword(java.lang.String,java.lang.String)! No property email found for type UserEntity! Did you mean 'u_email'?

위와 같습니다.

어차피 UserRepository에서 하는 일은 쿼리를 작성하는 일이고, 변수 명은 코드를 작성하는 사람 재량인데 왜 오류가 나는지 궁금합니다

**Comment by hkjs96** [comment, marked as answer]
https://docs.jboss.org/hibernate/orm/5.3/userguide/html_single/Hibernate_User_Guide.html#naming

https://www.jpa-buddy.com/blog/hibernate-naming-strategies-jpa-specification-vs-springboot-opinionation/

https://docs.jboss.org/hibernate/orm/5.4/javadocs/org/hibernate/boot/model/naming/Identifier.html


더 파보고 싶었으나,,,,

hibernate ORM 에 이름 명명 전략이라고 있다는데 까보니까

![image](https://user-images.githubusercontent.com/75015048/163957906-ba18ac20-98d1-4f9a-b313-f87fc0a57be6.png)

![image](https://user-images.githubusercontent.com/75015048/163957609-7fac13a9-ca08-41f6-99b1-5fe58429c512.png)

StringUtil.splitByCharacterTypeCamelCase() 라는 메서드에서 이름을 분리해주고 이친구가 CamelCase 로 변환해주는데
이름 명명의 기본 전략이 CamelCase -> SnakeCase 로 변환되기 때문이지 않을까 싶습니다???
전략을 바꿔주던지

@Column 으로 명명 해주시는게,,,, 
잘 모르겠네요.... 짬난시간에 찾아본거라
